lib/plugins/book.js

```diff
@@ -28,44 +31,44 @@ function inject (bot, { version }) {
     }
   }
 
-  function write (slot, pages, author, title, signing, cb = noop) {
+  async function write (slot, pages, author, title, signing) {
     assert.ok(slot >= 0 && slot <= 44, 'slot out of inventory range')
     const book = bot.inventory.slots[slot]
     assert.ok(book && book.type === mcData.itemsByName.writable_book.id, `no book found in slot ${slot}`)
     const quickBarSlot = bot.quickBarSlot
 
     if (slot < 36) {
-      bot.moveSlotItem(slot, 36, (err) => {
-        if (err) return cb(err)
-        bot.setQuickBarSlot(0)
-        modifyBook(36, pages, author, title, signing, (book) => {
-          function onSetInvSlot (oldItem, newItem) {
-            if (newItem.slot !== slot) return
-            bot.removeListener('setSlot:0', onSetInvSlot)
-            bot.setQuickBarSlot(quickBarSlot)
-            bot.moveSlotItem(36, slot, cb)
-            cb()
-          }
-          bot.on('setSlot:0', onSetInvSlot)
-          editBook(book, signing)
-        })
+      await bot.moveSlotItem(slot, 36)
+      bot.setQuickBarSlot(0)
+      const book = await modifyBook(36, pages, author, title, signing)
+      return new Promise(resolve => {
+        async function onSetInvSlot (oldItem, newItem) {
+          if (newItem.slot !== slot) return
+          bot.removeListener('setSlot:0', onSetInvSlot)
+          bot.setQuickBarSlot(quickBarSlot)
+          await bot.moveSlotItem(36, slot)
+          resolve()
+        }
+        bot.on('setSlot:0', onSetInvSlot)
+        editBook(book, signing)
```

Comment:
```js
// at the top of the file
const { once } = require('events')

editBook(book, signing)
let oldItem
let newItem
do {
  ;[oldItem, newItem] = await once(bot, 'setSlot:0')
} while (newItem.slot !== slot)

bot.setQuickBarSlot(quickBarSlot)
await bot.moveSlotItem(36, slot)
```

Reply:
This one introduces a race condition as well since the listener should be registered before executing the action

Reply:
Same as https://github.com/PrismarineJS/mineflayer/pull/1475#discussion_r534322910
But again you can write this like

```js
// at the top of the file
const { once } = require('events')

const response = once(bot, 'setSlot:0')
editBook(book, signing)
let [oldItem, newItem] = await response
while (newItem.slot !== slot) {
  ;[oldItem, newItem] = await once(bot, 'setSlot:0')
}

bot.setQuickBarSlot(quickBarSlot)
await bot.moveSlotItem(36, slot)
```